README.rst
pytypo
===========
|travis| |coveralls| |landscape| |pyversion| |version| |license|

pytypo corrects English spelling mistakes.
That feature is based on TYPO CORPUS (http://luululu.com/tweet/)
and Wikipedia (https://en.wikipedia.org/wiki/Wikipedia:Lists_of_common_misspellings/For_machines)

And this module normalizes also lengthened English expression having repeating letters.
(e.g., this module converts "cooooooooooooooollllllllllllll" to "cool")

That feature is based on the following paper:
Samuel Brody and Nicholas Diakopoulos.
Cooooooooooooooollllllllllllll!!!!!!!!!!!!!! using word lengthening to detect sentiment in microblogs.
In EMNLP2011, pp. 562-570, 2011.
http://aclweb.org/anthology//D/D11/D11-1052.pdf


Contributions are welcome!


Installation
============

::

 $ pip install pytypo


Usage
=====

Import pytypo
--------------------------------------------

::

 >>> import pytypo


correct sentence
--------------------------------------------

::

 >>> pytypo.correct_sentence('you are coooolll!!!')
 you are cool!


- correct_sentence(str)


correct word
--------------------------------------------

::

 >>> pytypo.correct('okayyyyy')
 okay


- correct(str)


Shorten repeated substring until threshould without dictionary
-------------------------------------------------------------------

::

 >>> pytypo.cut_repeat('mamisaaaaaan', 1)
 mamisan
 >>> pytypo.cut_repeat('okayyyyy', 2)
 okayy


- cut_repeat(str, threshould)

  * Note that this method don't use a lengthened expression normalize table (e.g., cooll -> cool).
    If you want to normalize such expression, use `correct()` or `correct_sentence()` method.

License
=========

- This module is licensed under MIT License.

.. |travis| image:: https://travis-ci.org/ikegami-yukino/pytypo.svg?branch=master
    :target: https://travis-ci.org/ikegami-yukino/pytypo
    :alt: travis-ci.org

.. |coveralls| image:: https://coveralls.io/repos/ikegami-yukino/pytypo/badge.svg?branch=master&service=github
    :target: https://coveralls.io/github/ikegami-yukino/pytypo?branch=master
    :alt: coveralls.io

.. |landscape| image:: https://landscape.io/github/ikegami-yukino/pytypo/master/landscape.svg?style=flat
   :target: https://landscape.io/github/ikegami-yukino/pytypo/master
   :alt: Code Health

.. |pyversion| image:: https://img.shields.io/pypi/pyversions/pytypo.svg

.. |version| image:: https://img.shields.io/pypi/v/pytypo.svg
    :target: http://pypi.python.org/pypi/pytypo/
    :alt: latest version

.. |license| image:: https://img.shields.io/pypi/l/pytypo.svg
    :target: http://pypi.python.org/pypi/pytypo/
    :alt: license


CHANGES.rst

CHANGES
==========

0.3 (2017-10-18)
---------------------

Add many cases from Wikipedia


0.2 (2016-04-15)
---------------------

Add many cases

0.1 (2016-04-14)
---------------------

First release.



test_pytypo.py
# -*- coding: utf-8 -*-
from nose.tools import assert_equal
import pytypo


def test_cut_repeat():
    assert_equal(pytypo.cut_repeat('pytypooooooo', 1), 'pytypo')
    assert_equal(pytypo.cut_repeat('beeeeer', 2), 'beer')


def test_correct():
    # TYPO CORPUS
    assert_equal(pytypo.correct('twiter'), 'twitter')
    # estimate by Wikipedia article titles
    assert_equal(pytypo.correct('speling'), 'spelling')
    # lengthned words
    assert_equal(pytypo.correct('thanxxxx'), 'thanx')
    assert_equal(pytypo.correct('coooooooolllll!!!!'), 'cool!')


def test_correct_sentence():
    normalized = pytypo.correct_sentence('you are coooolll!!!')
    assert_equal(normalized, 'you are cool!')


pytypo/pytypo.py
# -*- coding: utf-8 -*-
from sys import version_info
import re
from . import _conv_table

re_symbol = re.compile("[^a-zA-Z0-9']{2,256}$")

if version_info < (3, 0, 0):
    range = xrange


def cut_repeat(text, threshold):
    """Reduce repeated characters until threshold
    Param:
        <str> text
        <int> threshold
    Return:
        <str> result
    """
    text = list(text)
    result = text[0]
    count = 0
    for i in range(1, len(text)):
        if text[i - 1] == text[i]:
            count += 1
            if count < threshold:
                result += text[i]
        else:
            count = 0
            result += text[i]
    return result


def correct(word):
    """Normalize repeat expression for English
    Param:
        <str> word
    Return:
        <str> normalized_word
    """
    suffix_symbol = re_symbol.findall(word)
    if suffix_symbol:
        word = word.replace(suffix_symbol[0], '')
    word = cut_repeat(word, 2)
    normalized_word = _conv_table.eng_lengthened.get(word, word)
    if normalized_word == word:
        normalized_word = _conv_table.eng_typo.get(word, word)
    if suffix_symbol:
        return (normalized_word + cut_repeat(suffix_symbol[0], 1))
    return normalized_word


def correct_sentence(sentence):
    """Normalize for each word
    Param:
        <str> sentence
    Return:
        <str> normalized_sentence
    """
    return ' '.join([correct(word) for word in sentence.split()])


pytypo/estimate.py
# -*- coding: utf-8 -*-
# This code is based on http://norvig.com/spell-correct.html
import bz2
import json
import os
import pkgutil

alphabet = 'abcdefghijklmnopqrstuvwxyz'
voca = json.loads(pkgutil.get_data('pytypo', os.path.join('voca.json')).decode('utf8'))


def edits1(word):
    splits = [(word[:i], word[i:]) for i in range(len(word) + 1)]
    deletes = [a + b[1:] for a, b in splits if b]
    transposes = [a + b[1] + b[0] + b[2:] for a, b in splits if len(b) > 1]
    replaces = [a + c + b[1:] for a, b in splits for c in alphabet if b]
    inserts = [a + c + b for a, b in splits for c in alphabet]
    return set(deletes + transposes + replaces + inserts)


def known_edits2(word):
    return set(e2 for e1 in edits1(word) for e2 in edits1(e1) if e2 in voca)


def known(words):
    return set(w for w in words if w in voca)


def estimate(word):
    candidates = (known([word]) or known(edits1(word))
            or known_edits2(word) or [word])
    return max(candidates, key=voca.get)


pytypo/__init__.py
# -*- coding: utf-8 -*-
"""pytypo - Spell correction

This module corrects spelling mistake.
That feature is based on TYPO CORPUS (http://luululu.com/tweet/)
and Wikipedia (https://en.wikipedia.org/wiki/Wikipedia:Lists_of_common_misspellings/For_machines)

Also this module normalizes lengthened English expression having repeating letters.
(e.g., this module converts "cooooooooooooooollllllllllllll" to "cool")

That feature is based on the following paper:
Samuel Brody and Nicholas Diakopoulos.
Cooooooooooooooollllllllllllll!!!!!!!!!!!!!! using word lengthening to detect sentiment in microblogs.
In EMNLP2011, pp. 562-570, 2011.
http://aclweb.org/anthology//D/D11/D11-1052.pdf


Author:
    Yukino Ikegami

Lisence:
    MIT License

Usage:
    import pytypo
    text = pytypo.correct(text) # Normalization for a word
    text = pytypo.normalize_sentence(text) # Normalization for a sentence
"""
from . import pytypo

VERSION = (0, 3, 0)
__version__ = '0.3.0'
__all__ = ['correct', 'correct_sentence', 'cut_repeat']

correct = pytypo.correct
correct_sentence = pytypo.correct_sentence
cut_repeat = pytypo.cut_repeat
